[Uniformity] Fixed control-div early stop

jgu222 opened the pull request:
Control-divergence finds joins by propagating labels from the divergent control branch. The code that checks the early stop for propagation is not correct in some cases.

This PR, also included changes from ssahasra, fixes this issue by stopping no early than the post-dominator of  the divergent branch.

https://github.com/llvm/llvm-project/issues/137277

jgu222 committed:
[Uniformity] Fixed control-div early stop

Control-divergence finds joins by propagating labels from the divergent
control branch. The code that checks the early stop for propagation is
not correct in some cases.

This change fixes this issue by stopping at the post-dominator of
the successors of the divergent branch.

https://github.com/llvm/llvm-project/issues/137277

ssahasra commented:
I agree that #138806 should be discarded. I was able to construct an example which fails with that change. I was about to post that example, but then I saw this PR.

But what is the reasoning about the irreducible cycle? There are four cases:
1. `DivTermBlock` is in an irreducible cycle that does not contain the candidate join.
2. The candidate join is in an irreducible cycle that does not contain `DivTermBlock`.
3. Both are inside an irreducible cycle.

Case 3 is a subset of both case 1 and case 2. Clearly, in case 3 we should stop when `FreshLabels.size() == 1`, because we have found the IPD? What is exactly going on with case 2? Is it possible that sometimes inside an irreducible cycle, `FreshLabels.size() == 1` actually indicates that we found the IPD? All this seems to imply some inefficiency traversing unnecessary blocks.

I am somewhat concerned that we discovered a glaring bug in the early stopping criterion, and would prefer to be way more cautious when we are revisiting. Note that the IPD is the exact place where the propagation of labels should stop, because that is the last join point of `DivTermBlock`. Stopping at a later point makes the algorithm inefficient. Stopping at an earlier point is incorrect. Then essentially, this early stopping criterion is a roundabout way of discovering the IPD. Then why not just use the IPD? It is guaranteed to be correct, and I don't see how it will be any costlier than us computing the IPD using this traversal. Our traversal is performed once for every divergent branch, and it revisits nested regions. The IPD analysis is likely to traverse the whole graph once and should be cheaper.

jgu222 commented:
I agree that IPD is the last join and the algo should stop at the IPD.

I did experiment disabling the early stop, and I got some regressions on irreducible lit tests. I am not sure what is wrong and think maybe irreducible cycles are specially handled and rely on particular propagation.

As you think the IPD should be good for irreducible cycles too, I will redo the patch to use dominator analysis.

jgu222 commented:
@ssahasra  Given the following:
` 
; RUN: opt %s -mtriple amdgcn-- -passes='print<uniformity>' -disable-output 2>&1 | FileCheck %s

define amdgpu_kernel void @cycle_inner_ipd(i32 %n, i32 %a, i32 %b) #0 {
;
;          entry(6)
;        /      \
;     E2(5)<----E1(1)
;       | \     ^^
;       |  \  /  |
;       |   A(4) |
;       |  /     |
;       | /      |
;     B(3) ----->C(2)
;                |
;                X(0)
;
;
;
entry:
  %tid = call i32 @llvm.amdgcn.workitem.id.x()
  %div.cond = icmp slt i32 %tid, 0
  %uni.cond = icmp slt i32 %a, 0
  %uni.cond1 = icmp slt i32 %a, 2
  %uni.cond2 = icmp slt i32 %a, 10
  br i1 %uni.cond, label %E2, label %E1

E1:
  br label %E2

E2:
  br i1 %uni.cond1, label %A, label %B


A:
  br i1 %div.cond, label %E1, label %B

B:
  %div.merge = phi i32 [ 0, %A ], [ %b, %E2 ]
  br label %C

C:
  br i1 %uni.cond2, label %E1, label %X

X:
  ret void
}
'

jgu222 commented:
@ssahasra I have a new PR based on IPD https://github.com/llvm/llvm-project/pull/140013
However, it has three lit failures on irreducible tests.

ssahasra commented:
> Note that the IPD is the exact place where the propagation of labels should stop, because that is the last join point of `DivTermBlock`.

From the new testcase above, I believe this statement is wrong. E1 is a join that is reachable from A using a path that passes through B (the IPD) and C. So we cannot simply stop at IPD as in #140013 , and we cannot always use it as a floor either.

Falling back to first principles, the overall idea is that we have an ordering provided by ModifiedPO, and starting from the divergent branch, we traverse that order to find every join point. If the IPD is not inside any cycle, then we do not have to look beyond the IPD to find all the joins. If the IPD is inside a cycle (doesn't matter if the branch is in the same cycle or not), then there can be a join that is discovered only after encountering the IPD. In this case, the cycle header is guaranteed to be earlier in ModifiedPO by design, and should be used as the floor. We don't have to follow any fresh label that comes after the header.

So I think the correct solution is to compute the floor as the minimum of of IPD as one argument and the header of the cycle that contains the IPD as the second argument. But which cycle? Is it the smallest cycle containing the IPD, or the largest cycle? Or the common parent of both the divergent branch and the IPD?

ssahasra commented:
What I have in mind is something like this:

```
      // Early stopping criterion
      int DivTermIdx = CyclePOT.getIndex(&DivTermBlock);
      const auto *ImmPDom = getIPDom(&DivTermBlock);
      int FloorIdx = CyclePOT.getIndex(ImmPDom);
      LLVM_DEBUG(dbgs() << "IPDom " << Context.print(ImmPDom)
                 << "with index " << FloorIdx << "\n");

      CycleT *CommonCycle = CI.getSmallestCommonCycle(&DivTermBlock, ImmPDom);
      if (CommonCycle)
        FloorIdx = std::min<int>(FloorIdx, CyclePOT.getIndex(CommonCycle->getHeader()));
      LLVM_DEBUG(dbgs() << "FloorIdx: " << FloorIdx << "\n");
```

Then eliminate all the logic for updating FloorIdx, because we already know the minimum, and do the usual traversal that visits all newly labelled blocks such that `BlockIdx < FloorIdx`.

ssahasra commented:
Continuing along the previous comment, now I am wondering if we even need the ModifiedPO ordering. As far as I can see, for a node N with an IPDom P, if there is a common cycle C containing N and P, then we need to propagate labels through all the nodes in C until a fixed point is reached, else if there is no C, then propagate labels up to P. In either case, a simple BFS is enough and ModifiedPO gives no advantage.

jgu222 commented:
This changes seems handling the cases that IPD cannot handle.  With your comments, I looked into it further and conclude that 

> > Note that the IPD is the exact place where the propagation of labels should stop, because that is the last join point of `DivTermBlock`.

This is a little tricky.  If the execution favors A->E1->E2 cycles,  ie, if there is any active lanes going to E1, the code will execute E1.  The execution goes from A to B only if all active lanes  at A goes to B.   If so,  E1 is not a join point.  This is more like a lit test under AMDGPU/irreducible/irreducible-2.ll::@natural_loop_two_backedges(), which has the following CFG:

```
;          entry
;            |
;            H<-|<-|
;            |    |   |
;            B->     |
;            |        |
;            C----->
;            |
;            X
```

B is DivTermBlock,  is ph in H uniform ?   If the execution will honor the inner loop first, the phi in H is uniform. If the execution at B favors C (outer loop),  then phi in H is divergent.

Unless we are sure the inner loop is always favored,  we have to treat phi in H as divergent.  This also means that IPD isn't necessarily convergence point.   If so,  IPD should not be used.   Is IPD the max-convergence ?

This patch (139667) does not assume IPD is a convergence point.  It will show E1 is a join if making this example a reducible CFG by removing entry->E2 edge.

jgu222 commented:
> Continuing along the previous comment, now I am wondering if we even need the ModifiedPO ordering. As far as I can see, for a node N with an IPDom P, if there is a common cycle C containing N and P, then we need to propagate labels through all the nodes in C until a fixed point is reached, else if there is no C, then propagate labels up to P. In either case, a simple BFS is enough and ModifiedPO gives no advantage.

It seems correct.  But It seems to me that we may choose to stay the current course without using IPD ?

ssahasra commented on llvm/include/llvm/ADT/GenericUniformityImpl.h:
```diff
@@ -626,14 +623,36 @@ template <typename ContextT> class DivergencePropagator {
+    auto IsInIrreducibleCycle = [this](const BlockT *B) {
+      for (const auto *Cycle = CI.getCycle(B); Cycle;
+           Cycle = Cycle->getParentCycle()) {
+        if (!Cycle->isReducible())
+          return true;
```
```suggestion
        // If everything is inside a reducible cycle, then look no further
        if (Cycle->isReducible()) && Cycle->contains(DivTermBlock))
          return false;
        if (!Cycle->isReducible())
          return true;
```
I have not tested this thoroughly yet, but I believe we can stop early if we reach a reducible cycle that contains both the `DivTermBlock` and the current block. This ensures that we don't end up unnecessarily traversing a large cycle like in this situation:

```
C1 (irreducible) contains:
  C2 (reducible) contains:
     Divergent branch B and all possible join points
```
In this case, we should not continue traversing all of `C1` when only one fresh label is being reported. It needs to be proven, but intuitively, C2 is the cycle that contains both the branch and its IPD.

I am still thinking whether we should traverse `C1` or not if `C2` is irreducible.

jgu222 replied:
> I have not tested this thoroughly yet, but I believe we can stop early if we reach a reducible cycle that contains both the `DivTermBlock` and the current block. This ensures that we don't end up unnecessarily traversing a large cycle like in this situation:
> 

I updated the patch based on your suggestion. All tests passed for me.

ssahasra replied:
"All tests passed" is not a sufficient criterion for deciding that this heuristic actually works. I do believe that this is a sufficient work around for now, but a formal proof is pending. Let's go with this for now, but separately we still need to prove it beyond all doubt.

Also, in its current state, it may cause unnecessary traversal when the current block is in an irreducible cycle that does not contain `DivTermBlock`, and it is reachable only after the IPD of `DivTermBlock`. But this is rare and performance issue rather than a correctness issue.

jgu222 replied:
The algo may be conservative for some irreducible CFGs,  but it should be correct based on our convergence assumption.  I agree we need a formal proof.

ssahasra reviewed (commented):
@jgu222 , I see your point now. See comment about checking for outermost irreducible cycle.

jgu222 commented:
Thanks. I will come out a modified patch based on the feedback early next week.

jgu222 committed:
[UniformAnalysis] Changes based on feedback

Based on feedback, modifying early stop involving loops and minor lit test fix.

jgu222 commented:
With the latest patch, all lit tests passed.
I am  closing the other patch (https://github.com/llvm/llvm-project/pull/140013)

ssahasra committed:
[LLVM] [NFC] Add more tests for uniformity analysis

jgu222 committed:
Fix lit test based on feedback

jgu222 commented on llvm/test/Analysis/UniformityAnalysis/AMDGPU/irreducible/diverged-entry-headers-nested.ll:
```diff
@@ -126,10 +126,10 @@ exit:
-;; CHECK:   depth=1: entries(B A) D T S Q P R C
+;; NOCHECK: CYCLES ASSSUMED DIVERGENT:
+;; NOCHECK:   depth=2: entries(T P) S Q R
+;; NOCHECK: CYCLES WITH DIVERGENT EXIT:
```
fixed

jgu222 commented on llvm/test/Analysis/UniformityAnalysis/AMDGPU/irreducible/diverged-entry-headers-nested.ll:
```diff
@@ -126,10 +126,10 @@ exit:
-;; CHECK: CYCLES WITH DIVERGENT EXIT:
-;; CHECK:   depth=1: entries(B A) D T S Q P R C
+;; NOCHECK: CYCLES ASSSUMED DIVERGENT:
+;; NOCHECK:   depth=2: entries(T P) S Q R
```
fixed

jgu222 commented on llvm/test/Analysis/UniformityAnalysis/AMDGPU/irreducible/diverged-entry-headers-nested.ll:
```diff
@@ -126,10 +126,10 @@ exit:
-;; CHECK:   depth=2: entries(T P) S Q R
-;; CHECK: CYCLES WITH DIVERGENT EXIT:
-;; CHECK:   depth=1: entries(B A) D T S Q P R C
+;; NOCHECK: CYCLES ASSSUMED DIVERGENT:
```
fixed

ssahasra committed:
simpler logic for irreducible parent cycle

jgu222 committed:
Runs clang-format
